### PR TITLE
Dynamic memory allocation

### PR DESCRIPTION
**Pull request by mamboelrey**
Added code to create 3 new UDTs and relevant dynamic memory allocation techniques

### CODE REVIEW

**Review by matkatmusic — CHANGES_REQUESTED** [review]
1) please put the star and & next to the type name when you're declaring variables and return it types. 
2) you skipped this instructions:
```
 4) write add/subtract/divide/multiply member functions for each type that take your 3 UDTs
        These are in addition to your member functions that take primitives
        for example, IntType::divide(const DoubleType& dt);
        These functions should return the result of calling the function that takes the primitive.
```
specifically:
```
These functions should return the result of calling the function that takes the primitive.
```
3) you don't need to use the `this` keyword.

I am not going to mark every spot that needs revising.  Essentially every added function needs fixing, though.
- [x] revise

**Comment by mamboelrey** [comment]
Made required changes

**Review by mamboelrey — COMMENTED** [review]
Made the required changes

File: main.cpp [hunk 1]
```diff
@@ -112,325 +112,382 @@ struct DoubleType;
...
-    if(*(ft.value) == 0)
+    if(*it.value == 0)
     {
         std::cout << "error: integer division by zero is an error and will crash the program!"<< std::endl;
         return *this;
     }
-    *(this->value) /= *(ft.value);
-    return *this;
+    
+    return divide(*it.value);
```
Comment (matkatmusic) [inline main.cpp:245]: your `IntType::divide(int)` already has this std::cout message in it.
You don't need to duplicate the message here
all you need to do is call `return divide(*it.value);` in this function.
- [ ] revise

File: main.cpp [hunk 2]
```diff
@@ -112,325 +112,382 @@ struct DoubleType;
...
-    if(*(dt.value) == 0)
+    if(*ft.value == 0.0f)
     {
         std::cout << "error: integer division by zero is an error and will crash the program!"<< std::endl;
         return *this;
     }
     
-    *(this->value) /= *(dt.value);
-    return *this;
+    return divide(*ft.value);
```
Comment (matkatmusic) [inline main.cpp:268]: - [ ] revise, see above comments

File: main.cpp [hunk 3]
```diff
@@ -112,325 +112,382 @@ struct DoubleType;
...
+IntType& IntType::divide(const DoubleType& dt)
+{
+    if(*dt.value == 0.0)
+    {
+        std::cout << "error: integer division by zero is an error and will crash the program!"<< std::endl;
+        return *this;
+    }
+    
+    return divide(*dt.value);
+}
```
Comment (matkatmusic) [inline main.cpp:292]: - [ ] revise, see above comments

File: main.cpp [hunk 4]
```diff
@@ -112,325 +112,382 @@ struct DoubleType;
...
+    return multiply(*ft.value);
 }
-DoubleType &DoubleType::divide(const FloatType& ft)
+DoubleType& DoubleType::divide(const FloatType& ft)
 {
-    if(*(ft.value) == 0.0)
+    if(*ft.value == 0.0f)
     {
         std::cout << "warning: floating point division by zero!"<< std::endl;
     }
```
Comment (matkatmusic) [inline main.cpp:342]: - [ ] revise, see above comments

File: main.cpp [hunk 5]
```diff
@@ -112,325 +112,382 @@ struct DoubleType;
...
-DoubleType &DoubleType::divide(const IntType& it)
+DoubleType& DoubleType::divide(const IntType& it)
 {
-    *(this->value) /= *(it.value);
-    return *this;
+    if(*it.value == 0)
+    {
+        std::cout << "warning: floating point division by zero!"<< std::endl;
+    }
+    return divide(*it.value);
```
Comment (matkatmusic) [inline main.cpp:366]: - [ ] revise, see above comments

File: main.cpp [hunk 6]
```diff
@@ -112,325 +112,382 @@ struct DoubleType;
...
+    return multiply(*dt.value);
+}
+DoubleType& DoubleType::divide(const DoubleType& dt)
 {
-    *(this->value) += num;
+    if(*dt.value == 0.0)
+    {
+        std::cout << "warning: floating point division by zero!"<< std::endl;
+    }
+    return divide(*dt.value);
```
Comment (matkatmusic) [inline main.cpp:389]: - [ ] revise, see above comments

File: main.cpp [hunk 7]
```diff
@@ -112,325 +112,382 @@ struct DoubleType;
...
+    return multiply(*it.value);
 }
-FloatType &FloatType::divide(const IntType& it)
+FloatType& FloatType::divide(const IntType& it)
 {
-    if(*(it.value) == 0.0f)
+    if(*it.value == 0)
     {
         std::cout << "warning: floating point division by zero!"<< std::endl;
     }
```
Comment (matkatmusic) [inline main.cpp:443]: - [ ] revise, see above comments

File: main.cpp [hunk 8]
```diff
@@ -112,325 +112,382 @@ struct DoubleType;
...
-    if(*(dt.value) == 0.0f)
+    if(*ft.value == 0.0f)
     {
-        std::cout << std::endl << "warning: floating point division by zero!"<< std::endl;
+        std::cout << "warning: floating point division by zero!"<< std::endl;
     }
     
-    *(this->value) /= *(dt.value);
-    return *this;
+    return divide(*ft.value);
```
Comment (matkatmusic) [inline main.cpp:468]: - [ ] revise, see above comments

File: main.cpp [hunk 9]
```diff
@@ -112,325 +112,382 @@ struct DoubleType;
...
+    return multiply(*dt.value);
+}
+FloatType& FloatType::divide(const DoubleType& dt)
+{
+    if(*dt.value == 0.0)
+    {
+        std::cout << std::endl << "warning: floating point division by zero!"<< std::endl;
+    }
+
+    return divide(*dt.value);
```
Comment (matkatmusic) [inline main.cpp:490]: - [ ] revise, see above comments

**Review by matkatmusic — CHANGES_REQUESTED** [review]
D.R.Y.

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**